### ISSUE TITLE
Implement Continuous Integration

### ISSUE THREAD

**Issue by bobturneruk** [ScottishCovidResponse/SCRCIssueTracking#114]
Determine the best route to doing CI on this project GH Actions (used by CoronaBICI)? Travis? Appveyor? Also Codecov? Raise further issues for implementation.

**Comment by bobturneruk** [comment]
Chat on Zulip suggests GitHub actions is OK to start with. Now we have @aflag's excellent regression testing I suggest we:

- Set up GitHub actions
- Make it run when a pull request is made
- Make a successful run a criteria for merging to `master`, in addition to the current code review
- Make it use our conda environment for execution (ideally `spec-file.txt`, but `environment.yml` fine for now)
- Make it run `pytest`

A more complex example than I think we need is here https://help.github.com/en/actions/language-and-framework-guides/using-python-with-github-actions (ignore flake8 stuff in this e.g. for now, I think, also we don't need the test matrix as we have a single target environment)

**Comment by bobturneruk** [comment]
@aflag and @magicicada does this sound OK?

**Comment by aflag** [comment]
Sounds great. If we are also able to get a coverage report after the run it would be very nice too

**Comment by bobturneruk** [comment]
I think the coverage info. generated by pytest will be in the CI logs, could be pushed somewhere automatically. Have you any experience of using https://codecov.io ?

**Comment by aflag** [comment]
I've used https://coveralls.io/ before, never used codecov. But in jenkins there's a nice coverage UI that lets you look through the code and see what's covered and what's not. Having something like that is a nice to have, I think. But not really that critical. As you say, just looking at it in the pytest output is likely enough for now

**Comment by WPettersson** [comment]
Codecov also does let you drill down and show, line-by-line, which lines are run by tests, including how many times each line is encountered.

re. Jenkins, do you have a particular Jenkins host in mind?

**Comment by aflag** [comment]
i'm just talking about the futures I'd like, I'm not suggesting to use jenkins here. codecov sounds fine.

**Comment by WPettersson** [comment]
I've pushed a branch at https://github.com/ScottishCovidResponse/simple_network_sim/tree/issue-137-ci-on-travis that enables Travis CI testing and code coverage reports using Anaconda, feel free to build off it for getting this going.

**Comment by aflag** [comment]
Hm, we were talking about using github actions before, but given that you already have everything setup and working with travis, maybe we should just merge that in and consider it done, @bobturneruk ? Is anything missing from your implementation, @WPettersson ?

**Comment by bobturneruk** [comment]
Great to see this moving forwards! 

I don't object to Travis. From what I know they'll be doing fairly similar things. Almost certainly there's more Travis expertise in the world than in GitHub's equivalent. Maybe there's some accounts stuff to think about? Would we need a Travis (and codecov) account to get this working? I forget how all the triggering is set up...

Also, how is the coverage reported?

I note the current branch https://github.com/ScottishCovidResponse/simple_network_sim/blob/issue-137-ci-on-travis/.travis.yml uses `environment.yml`, not `spec-file.txt` which is more prescpiptive. Can this be changed? I'm guessing so.

Also, how hard would it be to switch to GitHub actions (effectively to port `.travis.yml`)?

**Comment by WPettersson** [comment]
I don't think anything is missing, but could be wrong. I also wrote it for Travis since that's what I'm most comfortable with (i.e. could do it off the top of my head). Happy to convert to Github Actions as well.

**Comment by alex-konovalov** [comment]
Just discovered this issue (thanks @aflag and @bobturneruk for pointing to it). Incidentally, I also submitted a PR https://github.com/ScottishCovidResponse/simple_network_sim/pull/24 adding support for Travis CI and Codecov. We use Travis CI and Codecov in a number of projects, that's why I choose them, but we also use Coveralls and experiment now with GitHub actions, and both seems very usable too. Since one could use them in parallel, I'd suggest to set up anything as soon as we can, to ensure that each commit and PR are tested, and then see how it goes and optimise CI setup as it goes.

**Comment by WPettersson** [comment]
So there's one outstanding issue at least, which is the choice between using `spec-file.txt` and `environment.yml`. I think using either is okay, and honestly I think it best to set up CI to test both. I also propose we use @alex-konovalov's direct python CI implementation (outside of Conda, and using python3.{6,7,8}). The `spec-file.txt` tests could be considered the de facto standard for making sure tests run, but the other tests would be useful to see how the code performs with `environment.yml` instead of `spec-file.txt`, or outside of conda completely, but failures would be allowed.

Anyone have any comments/issues regarding that plan?

**Comment by aflag** [comment]
I agree with you. If the spec-file.txt tests are passing we have confidence that we can recreate the same environment and get things working at any point in time in future. With environment.yml we are less certain, because it doesn't tell us all the versions of the eggs we need installed. But it's nice to test against environment.yml as well (specially if we can use it to test the code in mac and windows too), not passing there is not a blocker, but it's probably something we want to look into (there are contributors using all 3 OSes).

**Comment by WPettersson** [comment]
I pushed to the branch  issue-137-ci-on-travis recently - it's ready for merging, with a few caveats:

* I am also adding requirements.txt and requirements-test.txt for the
  extra tests
* I had to remove a line "@EXPLICIT" from spec-file.txt. I don't know
  what the purpose of this line is, it seemed extraneous and conda
  complained when trying to create an environment with it. If anyone
  more knowledgeable about conda can chime in, that'd be nice.
* Before any tests actually run, this repository will need to be set up
  for both Travis CI and codecov. For each of these, an admin will need
  to visit travis-ci.org or codecov.io respectively, log in using Github
  credentials, and enable this repository. It should be a relatively
  straight-forward process, but I'm happy to help people through.
  Only once this has been done, will future commits run through the
  testing (i.e., tests will run on commits that happen _after_ Travis CI
  is enabled).

Especially given the last point, I'm happy to wait until tomorrow some time if people want to ask more questions regarding enabling CI.

**Comment by bobturneruk** [comment]
I've asked @aflag to pick this up and progress to the point where we have some level of CI operational for now. Once this is up and going, it would be great to raise issues stemming from this thread to improve e.g. support running in multiple environments / OSs / Python versions etc. I think the most important things is to have an environment that's as tightly specified as we can get and runs the code consistently such that we can reproduce results exactly and on demand.

**Comment by bobturneruk** [comment]
We will need to ensure that any credentials are managed appropriately. Long term it's not appropriate for our CI to be linked to specific user account credentials, but we can work on this.

**Comment by bobturneruk** [comment]
Thank you for your excellent input on this topic - I really appreciate the discussion - it will definitely raise the standard of the project! Keep it coming!

**Comment by WPettersson** [comment]
On the credentials side, the CI won't be linked to specific user accounts. Instead, the CI requires someone with "admin" privileges (on Github) to enable CI and code coverage reports. Once it's enabled (by any admin), it's enabled for the repository without being linked to any one account.

Also, both Travis CI and codecov authenticate via Github, so there's no need for new accounts or anything like that either.

**Comment by bobturneruk** [comment]
Great @WPettersson - thanks for clearing that up.

**Comment by aflag** [comment]
@WPettersson regarding the requirements.txt and spec-file.txt. Didn't this work when creating the conda environment?

```conda create --name simple_network_sim --file spec-file.txt```

I'd rather not have two more files with the dependencies information. The reason we don't have a setup.py was that, for instance. And, if we do, we'll probably want to read the dependencies from the conda enviornment files.

**Comment by alex-konovalov** [comment]
Just to add that one can enable Travis and Codecov prior to merging the `issue-137-ci-on-travis` branch, and if a PR will be submitted from that branch, we will be able to check once more that the setup works prior to merging it.

**Comment by bobturneruk** [comment]
Thanks - I most recently used GitHub actions (happy with Travis though) thus got myself confused. I think @aflag will implement this if sufficiently "privileged"!

**Comment by WPettersson** [comment]
@aflag That conda script works fine (bar the possibly-extraneous @EXPLICIT tag) for the conda tests, but the regular python tests need some other way of reading the requirements. If we don't want to store the requirements*.txt files, one option is to instead store some way of extracting pip-suitable information from environment.yml. The other option is to drop the tests from outside of conda, so we only test the package from inside conda.

**Comment by aflag** [comment]
Which regular python tests? The current test requirements are already inside environment.yml, aren't they?

BTW, this seems to work fine without complaints here:

```
rafael@rafael-XPS-13-7390:~/PycharmProjects/simple_network_sim$ conda create --name test --file spec-file.txt 
Preparing transaction: done
Verifying transaction: done
Executing transaction: done
rafael@rafael-XPS-13-7390:~/PycharmProjects/simple_network_sim$ 
```

what's the command did you use exactly to read the spec-file.txt?

**Comment by WPettersson** [comment]
For regular python tests, I mean not using conda, but just python directly.

For the spec-file.txt issue, see https://travis-ci.com/github/WPettersson/simple_network_sim/jobs/336185665 - it complains about @explicit but if I remove that line, it works fine. It's downloading the latest conda on each run, so it can't be failing due to an old version, and I don't know what else it could be.

**Comment by aflag** [comment]
@WPettersson I think the idea is that conda will always be used. The command being executed in that job is ```conda env create --name simple_network_sim -f spec-file.txt```. Note that that's slightly different from ```conda create --name test --file spec-file.txt```

**Comment by WPettersson** [comment]
Ooh, thanks for that pointer. I didn't realise that the env was even missing. I wonder what the technical difference actually is.

I'll go back to just testing within Conda. Do you still want two tests, one with spec-file.txt and one with environment.yml? Or just the one test, with spec-file.txt?

**Comment by aflag** [comment]
I think we can start off with the spec-file.txt only.

**Comment by WPettersson** [comment]
Ok I've made a PR at https://github.com/ScottishCovidResponse/simple_network_sim/pull/26 for CI using the spec-file.

**Comment by bobturneruk** [comment]
Current plan is 100% conda. Happy to discuss this, but maybe needs it's own issue.

**Comment by WPettersson** [comment]
Closed in https://github.com/ScottishCovidResponse/simple_network_sim/pull/26
